src/client/dispatch.rs

```diff
@@ -17,18 +17,34 @@ use crate::model::guild::Member;
 use crate::model::id::GuildId;
 
 #[cfg(feature = "cache")]
-fn update_cache<E: CacheUpdate>(context: &Context, event: &mut E) -> Option<E::Output> {
-    context.cache.update(event)
+macro_rules! if_cache {
+    ($e:expr) => {
+        $e
+    };
 }
 
 #[cfg(not(feature = "cache"))]
-fn update_cache<E>(_: &Context, _: &mut E) -> Option<()> {
-    None
+macro_rules! if_cache {
+    ($e:expr) => {
+        None
+    };
+}
+
+#[cfg(feature = "cache")]
+macro_rules! update_cache {
+    ($cache:ident, $event:ident) => {
+        $event.update($cache)
+    };
+}
+
+#[cfg(not(feature = "cache"))]
+macro_rules! update_cache {
+    ($cache:ident, $event:ident) => {};
 }
 
 pub(crate) fn dispatch_model(
     event: Event,
-    context: Context,
+    context: &Context,
```

Comment:
```suggestion
    context: Context,
```
No need to change this I don't think.

Reply:
There is, otherwise you get clippy::needless_pass_by_value.